Replace CRA tooling with esbuild+Make

mariuswilms opened the pull request:
Building upon #140, I have our frontend now being build with esbuild and some Make, instead of using webpack and the tooling coming with CRA. This frees us from following CRA.

As with the CRA tooling `make dev` will also bring up a development server on port 3000, forwarding all API requests to a running backend instance on port 8080. The esbuild development server does not live reload, instead developers must refresh the browser which will cause a rebuild and reserver of the changed files. 

We should probably merge this PR after user-components have been merged, as I had to use the jsx suffix for all files containing JSX. This might create a great amount of conflicts otherwise.

wegry committed:
Rough draft of what the JS/HTML template building might look like

wegry committed:
Only parse JS/HTML templates on startup

mariuswilms committed:
WIP: Interactive Playgrounds

- Support the `-components` flag on `dsk`, to allow passing
  a path to a pre-build component bundle output directory.
- Provide playground handlers.
- Refactor `<Playground>`
- Use IFrame in Playground

Unrelated:

- Deprecate Node.components response property.
- Improve routing techniques, by using sub routers and allow
  us to extract route params.
- Rename flag variables.
- Pass more parent and current node into documentation components.
- Nodes now have an `id` that can be used to adress them. Although not
  from the outside.

mariuswilms committed:
Fix rebase

mariuswilms committed:
Use newly introduced file based templates

mariuswilms committed:
Fit both code brtanches together again

mariuswilms committed:
Bring back nodeLinker

wegry committed:
Vend yarn 3

wegry committed:
Patch up client GitHub actions pathing

mariuswilms committed:
WIP

mariuswilms committed:
WIP

mariuswilms committed:
WIP

mariuswilms committed:
Seperate IndexJS handler

mariuswilms committed:
Unbreak package.json

mariuswilms committed:
Unbreak package.json

mariuswilms committed:
clean

mariuswilms committed:
Unbreak

mariuswilms committed:
Use matching dev branch

wegry committed:
Running built playground iframe, now onwards to the layout

wegry committed:
Remove local module resolution

wegry committed:
Patch up lock file

wegry committed:
Load custom css index and ham in custom component rendering

christophlocher committed:
Run prettier

mariuswilms committed:
Merge branch '1.4' into user-components

mariuswilms committed:
Merge branch '1.4' into user-components

wegry committed:
Try a few different css entry point names

christophlocher committed:
Frontend WIP

christophlocher committed:
Frontend Refactoring – Remove useGlobal, switch to React Router

christophlocher committed:
Make sure Id for NodeDocComponent is unique

Before, this generated the same Id for any component that had the same content no matter the tag or attributes, which leads to problems with empty components: <ColorCard color="#FF00000"></ColorCard> and <TableOfContents></TableOfContents> got the same Id. Adding the position to the string to be hashed ensures uniqueness.

christophlocher committed:
Fixes

wegry committed:
Patch up lock file

wegry committed:
Look up js entry point based on esbuild module resolution

wegry committed:
Update contributing example component directory path

wegry committed:
Use automatic jsx transform that's part of CRA@>4

https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html\#create-react-app

wegry committed:
Get front end tests running with react 17

wegry commented on frontend/esbuild.js:
```diff
@@ -0,0 +1,68 @@
+      console.log('(Expecting backend at localhost:8000 for API request routing.)');
+    });
+} else {
+  esbuild.build({ ...bundle, color: true, logLevel: 'debug' }).catch(() => process.exit(1));
```
Because this is async, line https://github.com/rundsk/dsk/pull/143/files#diff-43db6ee53c7d9300f7c10130f4d7186138e77b2e4299bab1de2bb350ae4f83e0R67 will fire immediately after (and probably before this actually builds). We can either use top level await (which has some gotchas) or

```suggestion
  try {
    esbuild.buildSync({ ...bundle, color: true, logLevel: 'debug' })
  } catch { 
     process.exit(1) 
  }
```

mariuswilms replied:
Thanks for pointing that out, unfortunatly `buildSync` is incompatible with the use of plugins. I've  got away with dropping the log message and using promise syntax.

wegry commented:
> The esbuild development server does not live reload, instead developers must refresh the browser which will cause a rebuild and reserver of the changed files.

@ChristophLabacher how do you feel about this? I ejected from CRA to fix front end tests #139 and rued the day. The CRA webpack config alone is enormous.

Maybe the tests don't matter though. And maybe live reload isn't a huge deal (I'll admit I'm very fond of it).

christophlocher committed:
Only add the source to a link when it is not explicitly set

Used to overwrite the version that was given in the link

christophlocher committed:
Styling

christophlocher committed:
Restructure frontend direcotries to modern React standards

It has become a standard to keep the component code in a file called Component.jsx and reexport the component in index.js.

wegry committed:
Use ESM instead of IIFE for playground format

wegry committed:
Target ES2020 with playgrounds

wegry committed:
Output paths are stable once again for playgrounds

christophlocher committed:
Add componets to docs in API, pass them to transform

Together with the “Treat content of components that preformatted as text” commit of js-sdk, this makes sure using html in preformatted components does not introduce strange characters in our output anymore.

christophlocher committed:
Fix API Routing for root node

christophlocher commented:
I would definitely prefer to get rid of the webpack config. I do like the live-reload a lot, but we could try without.

@mariuswilms I just renamed a lot of files in the `user-components` branch, so there will probably be a lot of conflicts. Sorry 😬 On the other hand I renamed them to `.jsx` …

mariuswilms commented:
I've rebased this onto `user-components`, and can merge this once that branch is merged.

mariuswilms commented:
Test now don't pass but run. Please note that they don't pass as there is an issue in user-components that makes the tests fail. It should be adressed there.

mariuswilms committed:
Fix linting

mariuswilms committed:
Format

mariuswilms committed:
Move runtime.jsx, use embed, rename vars

wegry commented:
The back or front end? I’d gotten the front end running with react-scripts. We can just roll our own jest config though.

> On Oct 14, 2021, at 09:39, Marius Wilms ***@***.***> wrote:
> 
> ﻿
> Test now don't pass but run. Please note that they don't pass as there is an issue in user-components that makes the tests fail. It should be adressed there.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

mariuswilms commented:
> The back or front end? I’d gotten the front end running with react-scripts. We can just roll our own jest config though.

I stand corrected it's an issue introduced in this branch. I've removed support for objects (React nodes) as children of CodeBlock. You receive those when you use the component (not the markdown fenced code syntax) and don't use script tags to wrap the inner content. I removed support for that in this branch as it required React DOM Server which in turn requires additional shims when build through esbuild.

It's a feature we don't need anyway and can live without, since Christoph figured how to ensure we always receive unparsed string content in CodeBlock.

I'll be fixing that test here. When everything looks good and user-components is merged into 1.4 I'll squash/separate out commits in this branch and merge them into 1.4.

mariuswilms committed:
Use temporary outdir

christophlocher committed:
CodeBlock content is now always escaped

christophlocher committed:
Heading styling improvements

christophlocher committed:
Add support for component CSS path „style.css“

christophlocher committed:
Add „Show Source“ Button to ReactPlaygrounds

mariuswilms committed:
Replace CRA with esbuild+Make

mariuswilms committed:
Unsupport react nodes in CodeBlock

This requires React DOM Server which when build through esbuild requires
additional shims.

mariuswilms committed:
Optimize bundle size, use less highlighting langs

mariuswilms committed:
Format

mariuswilms committed:
Fixup CodeBlock unsupporting

mariuswilms commented:
I've fixed the failing test in user-components, the commit that introduced the issue here belonged there. As user-components is where we got rid of the necessity of supporting react nodes in CodeBlock.

mariuswilms commented:
Last remaining bit here is fixing esbuild config to support data:image URLs in CSS, or finding a workaround for that.

mariuswilms committed:
Unbreak tests

mariuswilms committed:
Fix lint issue in Node.Id

mariuswilms committed:
Uninline SVG image

mariuswilms commented:
Also fixed that, all done, waiting for user-components to merge, or additional reviews that block this PR.

mariuswilms committed:
Remove ejected config/scripts
